chain/status.go

```diff
@@ -0,0 +1,169 @@
+package chain
+
+import (
+	"fmt"
+	"sync"
+
+	logging "github.com/ipfs/go-log"
+
+	"github.com/filecoin-project/go-filecoin/types"
+)
+
+var logChainStatus = logging.Logger("chain/status")
+
+// Reporter defines an interface to updating and reporting the status of the blockchain.
+type Reporter interface {
+	UpdateStatus(...StatusUpdates)
+	Status() Status
+}
+
+// StatusReporter implements the Reporter interface.
+type StatusReporter struct {
+	statusMu sync.Mutex
+	status   *Status
+}
+
+// UpdateStatus updates the status heald by StatusReporter.
+func (sr *StatusReporter) UpdateStatus(update ...StatusUpdates) {
+	sr.statusMu.Lock()
+	defer sr.statusMu.Unlock()
+	for _, u := range update {
+		u(sr.status)
+	}
+	logChainStatus.Debugf("syncing status: %s", sr.status.String())
+}
+
+// Status returns a copy of the current status.
+func (sr *StatusReporter) Status() Status {
+	return *sr.status
+}
+
+// NewStatusReporter initializes a new StatusReporter.
+func NewStatusReporter() *StatusReporter {
+	return &StatusReporter{
+		status: newDefaultChainStatus(),
+	}
+}
+
+// Status defines a structure used to represent the state of a chain store and syncer.
+type Status struct {
+	// The heaviest TipSet that has been fully validated.
+	ValidatedHead types.TipSetKey
+	// The height of ValidatedHead.
+	ValidatedHeadHeight uint64
+
+	// They head of the chain currently being fetched/validated, or undef if none.
+	SyncingHead types.TipSetKey
+	// The height of SyncingHead.
+	SyncingHeight uint64
+	// Whether SyncingTip is trusted as a head far away from the validated head.
+	SyncingTrusted bool
+	// Unix time at which syncing of chain at SyncingHead began, zero if valdation hasn't started.
+	SyncingStarted int64
+	// Whether SyncingHead has been validated.
+	SyncingComplete bool
+	// Whether SyncingHead has been fetched.
+	SyncingFetchComplete bool
+
+	// The key of the tipset currently being fetched
+	FetchingHead types.TipSetKey
+	// The height of FetchingHead
+	FetchingHeight uint64
```

Comment:
The head and height of the tipeset the fetcher is current;y fetching, these will change as the fetcher traverses from `SyncingHead` back until its done callback is satisfied.